learn/app-development/services/database-crud-event-listeners.md

```diff
@@ -0,0 +1,199 @@
+---
+title: "Database CRUD Operations Event Listeners"
+id: ""
+sidebar_label: "CRUD Event Listeners"
+---
+---
+
+When you create a new database or import an existing database in WaveMaker, the WaveMaker platform generates the database service inside the WaveMaker application.
+
+The database service code that gets generated inside the application is layered as shown below.
+
+
+## Layered Architecture
+
+1. Controller Layer
+	
+    The Controller Layer is responsible for the HTTP request and response processing, converting to and from JSON to Java Object, etc. The Controller layer code is annotated with the Spring REST API annotations.
+
+2. Service Layer
+
+    The Service layer is responsible for static validations, transactional boundaries and invokes the respective DAO methods for object persistence.
+
+3. DAO layer
+
+    The DAO layer is responsible for communicating with the database. It maintains a pool of connections with the database and uses Hibernate ORM for persistence.
+
+### Generated Code and Syncing Database
+
+The generated code is always in sync with the database schema. If any changes are made externally to the database schema, the respective database must be reimported, thus making the code regenerated to keep the code & schema in sync.
+
+:::note
+The platform manages the generated code, and therefore it does not allow any modifications to it or add the custom business logic. 
+:::
+
+### Database CRUD Operations
+
+In most cases, the application developers should have control on the database CRUD operations to perform the following scenarios:
+
+1. Dynamic Validations
+2. Custom Business Logic
+3. Authorization checks / Row-level security
+4. Encrypt or Decrypt the field values
+5. Set default/override values during insert/update
+6. Invoke additional services for integration
+7. Audit history for the data changes etc
+
+:::note
+The generated code is not editable to include the above use-cases when performing database operations.
+:::
+
+To address all the above use-cases, WaveMaker introduces Event Listeners for the typical database operations.
+
+## Listen and Handle Events
+
+WaveMaker’s application runtime framework publishes the entity CRUD event during the respective database operation. Each of the operations, aka Create, Read, Update and Delete, publishes two events - they are pre and post operation events. The WaveMaker application developer should listen to the “interested” events and handle the respective events in a Java Service.
+
+Each event extends [EntityCRUDEvent](https://github.com/wavemaker/wavemaker-app-runtime-services/blob/master/wavemaker-app-runtime-core/src/main/java/com/wavemaker/runtime/data/event/EntityCRUDEvent.java). These have methods to get the serviceId and entityClass for which the event is generated.
+
+The events can be listened to using the @EventListener or @TransactionalEventListener annotations.
```

Comment:
Please add the link for EventListener(org.springframework.context.event.EventListener) and TransactionEventListerer(org.springframework.transaction.event.TransactionEventListener)